rfcs/0013-transition-caching.md

```diff
@@ -0,0 +1,42 @@
+# Transition Caching
+
+## Summary
+[summary]: #summary
+
+A new transition caching logic for the transition router which aims to track transitions which are still being processed as well as transitions which have failed validation.
+
+## Motivation
+[motivation]: #motivation
+
+Within the transition router system, the only check for duplicate transitions is performed by the transition validator, and each transition is only checked against the transitions which are currently in the transition frontier. However, there are two types of duplicate transitions which are not being checked for: transitions which are still being processed by the system (either in the processor pipe or in the catchup scheduler and catchup thread), and transitions which have been determined to be invalid. In the case of the former, the system ends up processing more transitions than necessary, and the number of duplicated processing increases along with the networks size. In the case of the latter, the system is opened up for DDoS attacks since an adversary could continously send transitions with valid proofs but invalid staged ledger diffs, causing each node to spend a significant enough amount of time before invalidating the transition each time it recieves it.
```

Comment:
I don't think this is a viable attack. The attacker has to prove a blockchain snark in order to send you something that's going to get them immediately IP banned. And after they're banned we're not going to see that same transition again, so I don't see how a cache helps. Unless honest peers forward transitions before validating them? If that's the case we need to stop doing that.

Reply:
Oh, did you mean an attacker proves the snark once and then uses a botnet or something to send the associated transition many times? That's certainly more viable, if validating staged ledger diffs is expensive enough, though I'm very reluctant to add even more complexity to Coda.